File: VK_problem1.java
package javaapplication1;
import java.util.*;

class VK_problem1 {
    public int numOfIslands(int[][] grid){
        int row = grid.length;
        int col = grid[0].length;
        ArrayList<Integer> visit;
        visit = new ArrayList<>();
             
        boolean[][] dfs = null;
        int islands = 1;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(dfs[i][j]){
                    islands += 1;
                }
            }
        }
        return islands;
    }
    public boolean dfs(int grid[][], int r, int c, ArrayList<Integer> visit){ 
        int row = grid.length;
        int col = grid[0].length;

        if(r<0 || r>= row || c<0 || c>= col || grid[r][c] == "0" or visit.contains((r,c))){
            return false;
        }
        visit.add((r,c));
        if(dfs(r + 1, c)){
            return True;
        }
        if(dfs(r - 1, c)){
            return True;
        }
        if(dfs(r, c + 1)){
            return True;
        }
        if(dfs(r, c - 1)){
            return True;
        }
        visit.remove((r,c));
        
    }
    public static void main(String[] args) {
        JavaApplication1 obj1 = new JavaApplication1();
        obj1.Start();
    }
}

File: vk_problem1.py
# find the number of islands
# approach dfs + backtracking
def numOfIslands(grid):
    row = len(grid)
    col = len(grid[0])
    islands = 1
    visit = set()

    def dfs(r,c):
        if not(0<=r<row) or not(0<=c<col) or grid[r][c] == "0" or (r,c) in visit:
            return False
        visit.add((r,c))
        if dfs(r + 1, c): return True
        if dfs(r - 1, c): return True
        if dfs(r, c + 1): return True
        if dfs(r, c - 1): return True
        visit.remove((r,c))
        # removing so that in next iteration we can check the same grid element again

    for i in range(row):
        for j in range(col):
            if dfs(i,j):
                islands += 1
    return islands

grid= [ ["1", "0", "0", "1"],
        ["1", "1", "0", "0"],
        ["0", "1", "0", "0"],
        ["0", "0", "0", "0"] ]
print(numOfIslands(grid))

